### PR TITLE
[Feature][Diffusion] Plugin interface for third-party sparse attention for per-forward DiT

### PR DESCRIPTION
**Pull request by harenome**
## Purpose

Refactors and supersedes (#2231).
Implements a light plugin mechanism for sparse attention as proposed in RFCs #2233 and #3715.
It includes two parts:
- a generic per-forward state channel (to be forwarded to sparse attention kernels)
- a sparse attention dispatcher

This mechanism reuses and extends the ideas of per-role attention config (#2681) and `ForwardContext.denoise_step_idx` (#2640).

The dispatcher is built to make shipping a third-party sparse kernel as simple as possible. A plugin only has to **expose one callable** in its entry point and there is no base class to subclass or boilerplate: a minimal plugin can be just a function. Optional additional capabilities are advertised in a simple `capabilities` dict.

### Per-forward state (generic)

New field `AttentionMetadata.per_forward` holds a `PerForwardState` object for per-forward metadata.
It holds common metadata such as framework state or geometry.
Framework state (such as denoising steps) is read from `ForwardContext` by `Attention` layers.
Geometry is model-specific and set by the transformer forward of supported models.

### Sparse attention dispatcher

The sparse attention dispatcher lives in `SparseAttentionBackend` and selected via `DiffusionAttentionBackendEnum.SPARSE_ATTN`.
Plugins register a callable under the `vllm_omni.sparse_attn` entry-point group.

A minimal plugin can be as simple as the following snippet (on top of setting the entry-point):

```python
# A minimal plugin: a callable, plus an optional capabilities dict.
def my_attn(q, k, v, params, is_causal, softmax_scale) -> Tensor:
    ...  # return the attention output in BSND layout

# Optional: advertise framework inputs the kernel actually consumes.
my_attn.capabilities = {"supports_attention_mask": True}
```

It is enabled on per role with a short form: for instance, `per_role.self.backend=<plugin>` or `per_role.cross.backend=<plugin>` will be routed to the dispatcher automatically.
Multiple sparse plugins can coexist.
Built-in backends (other `DiffusionAttentionBackendEnum` values) have priority over third-party plugins.

The callable signature is `fn(query, key, value, params, is_causal, softmax_scale) -> Tensor`, with q/k/v in BSND layout `[batch, seq, heads, head_dim]`.

`params` is a flat dict (RFC #3715, Proposal B), merged from three sources in increasing priority:

1. static config — the kernel's own `extra.params`;
2. `PerForwardState` — the per-forward fields the framework fills (step, geometry);
3. `AttentionMetadata.extra` — dynamic per-forward values.

On a key collision the later source wins, and `None` values are dropped, so a kernel can rely on `params.get(key, default)`. 

Most `params` entries are optional and may be silently ignored by kernels.
Entries that can not be ignored, if present, are guarded by capabilities.
Currently, two entries must be handled if provided:
- `attn_mask` (for padding/validity, capability name: `supports_attention_mask`)
- `full_attn_spans` (for mixed causal/full attention, capability name: `supports_full_attn_spans`).

Additional capabilities can be advertised by setting an additional attribute `capabilities` to the callable.

## Ready for future evolutions

- Current demo examples are simple examples where `attn_mask` and `full_attn_spans` are `None`.
  The plugin mechanism is already ready for models and usecases that require it.
- New entries can be added in `params` as new usecases emerge and new entries will not break compatibility with older plugins
- `PerForwardState` uses per-sample tensors for denoising steps: ready for future work such as
  RFC #874 and #4024.

## Tests and docs

- `tests/diffusion/attention/test_per_forward_state.py`: the state dataclass and the layer helper (with a mocked SDPA backend).
- `tests/diffusion/sparse_attn/test_sparse_attn.py`: plugin lookup (mocked entry points), the three-layer merge order and `None`-drop, capability checks, and the short-form routing.

User docs: `docs/user_guide/diffusion/attention_backends.md` (the *Sparse attention (plugin)* section: the contract, `params`, capabilities, and a small example plugin).

## Demo

Two external plugins exercise the dispatcher end-to-end, each a standalone package with no dependency on vllm-omni: `spargeattn` (SpargeAttn) and `lvsa_sparse` (LVSA, with SDPA or FlashInfer kernels). See [vllm-omni-plugin-demo](https://github.com/harenome/vllm-omni-plugin-demo) for setup and run commands.

### CODE REVIEW

**Comment by Gaohan123** [comment]
@ZJY0516 PTAL

File: vllm_omni/diffusion/attention/layer.py [hunk 1]
```diff
@@ -240,6 +286,7 @@ def forward(
         query, key, value, attn_metadata, ctx = strategy.pre_attention(query, key, value, attn_metadata)
 
         attn_metadata = self._with_kv_cache_dtype(attn_metadata)
+        attn_metadata = self._with_per_forward_state(attn_metadata, query)
 
         # 2. Kernel Execution (Computation)
         if self.use_ring and strategy is not self._no_parallel_strategy:
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/attention/layer.py:370]: This bypasses the selected `SPARSE_ATTN` backend when `ring_degree > 1`: `RingParallelAttention` only runs its own SDPA/flash-style kernels, so a configured sparse plugin is silently ignored. Please either fail fast for sparse plugins with ring SP or route the ring path through the plugin, with a regression test.

File: vllm_omni/diffusion/attention/layer.py [hunk 2]
```diff
@@ -240,6 +286,7 @@ def forward(
         query, key, value, attn_metadata, ctx = strategy.pre_attention(query, key, value, attn_metadata)
 
         attn_metadata = self._with_kv_cache_dtype(attn_metadata)
+        attn_metadata = self._with_per_forward_state(attn_metadata, query)
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/attention/layer.py]: This now builds per-forward tensors for every attention layer whenever step/geometry exists, even for dense backends that ignore `per_forward`. Wan sets this state unconditionally, so default inference can pick up many extra GPU allocations per denoise step. Please gate this to backends that consume it, cache the state per forward, or include before/after latency evidence showing the default path is unaffected.

**Comment by Gaohan123** [comment]
@harenome Please resolve conlicts. Thanks
